server/service/src/invoice/outbound_return/generate_outbound_return_lines.rs

```diff
@@ -20,68 +24,222 @@ pub struct GenerateOutboundReturnLinesInput {
-    }: GenerateOutboundReturnLinesInput,
+    input: GenerateOutboundReturnLinesInput,
 ) -> Result<ListResult<OutboundReturnLine>, ListError> {
-    if stock_line_ids.is_empty() && item_id.is_none() && return_id.is_none() {
-        return Ok(ListResult {
-            count: 0,
-            rows: vec![],
-        });
-    }
+    // If a return_id is provided, get all existing lines for the provided stock_line_ids/item_id
+    let existing_return_lines = get_existing_return_lines(ctx, store_id, input.clone())?;
 
-    let mut filter = StockLineFilter::new();
+    // Get stock lines for any stock_line_ids passed in, regardless of whether that stock line is currently available
+    let from_stock_line_ids = stock_lines_from_stock_line_ids(ctx, store_id, input.stock_line_ids)?;
+
+    // If an item id is provided, get each stock line where stock is available
+    let from_item_id = stock_lines_for_item_id(ctx, store_id, &input.item_id)?;
+
+    let new_return_lines = vec![from_stock_line_ids, from_item_id]
+        .into_iter()
+        .flatten()
+        // filter out any stock lines for which we already have a return line (existing are joined below)
+        .filter(|new_line| {
+            !existing_return_lines.iter().any(|existing_line| {
+                new_line.stock_line_row.id == existing_line.stock_line.stock_line_row.id
+            })
+        })
+        .map(stock_line_to_new_return_line)
+        .collect::<Vec<OutboundReturnLine>>();
+
+    // return existing lines first, then new lines
+    let mut return_lines = existing_return_lines;
+    return_lines.extend(new_return_lines);
+
+    Ok(ListResult {
+        count: return_lines.len() as u32,
+        rows: return_lines,
+    })
+}
+
+fn stock_lines_from_stock_line_ids(
+    ctx: &ServiceContext,
+    store_id: &str,
+    stock_line_ids: Vec<String>,
+) -> Result<Vec<StockLine>, RepositoryError> {
+    let stock_line_repo = StockLineRepository::new(&ctx.connection);
+
+    let filter = StockLineFilter::new().id(EqualFilter::equal_any(stock_line_ids));
+
+    let stock_lines = stock_line_repo.query_by_filter(filter, Some(store_id.to_string()));
+
+    stock_lines
+}
+
+fn stock_lines_for_item_id(
+    ctx: &ServiceContext,
+    store_id: &str,
+    item_id: &Option<String>,
+) -> Result<Vec<StockLine>, RepositoryError> {
+    let stock_line_repo = StockLineRepository::new(&ctx.connection);
 
-    if !stock_line_ids.is_empty() {
-        filter.id = Some(EqualFilter::equal_any(
-            stock_line_ids.iter().map(String::clone).collect(),
-        ))
-    }
     match item_id {
         Some(item_id) => {
-            filter.item_id = Some(EqualFilter::equal_to(&item_id));
-            filter.is_available = Some(true);
+            let filter = StockLineFilter::new()
+                .item_id(EqualFilter::equal_to(item_id))
+                .is_available(true);
+
+            let stock_lines = stock_line_repo.query_by_filter(filter, Some(store_id.to_string()));
```

Comment:
Actually the store_id here is the filter for item visibility, do you know if stock lines are actually filtered for this store ? What if we add stock lines for another store would they be captured here ?